- Description: cio: Correct handling of no path during path verification.
  Symptom:     Linux hangs because path grouping is tried endlessly.
  Problem:     If a subchannel has lpm==0 and path verification is performed
               (as may happen after logical vary off), cio_start() will return
               -ENODEV (and not -EACCES). The subchannel's vpm is not adapted in
               that case and the while loop in __ccw_device_verify_start() will
               never finish because vpm and lpm will always be unequal.
  Solution:    Handle -ENODEV like -EACCES and switch off current path in the
               vpm.
  Problem-ID:  -
  Archs:       s390-31, s390-64

--- linux-2.5/drivers/s390/cio/device_pgid.c	19 Feb 2004 20:43:19 -0000	1.21
+++ linux-2.5/drivers/s390/cio/device_pgid.c	23 Apr 2004 14:40:37 -0000	1.21.2.1
@@ -227,7 +227,7 @@
 		ret = cio_start (sch, cdev->private->iccws,
 				 cdev->private->imask);
 		/* ret is 0, -EBUSY, -EACCES or -ENODEV */
-		if (ret != -EACCES)
+		if ((ret != -EACCES) && (ret != -ENODEV))
 			return ret;
 	}
 	/* PGID command failed on this path. Switch it off. */

